modules/trustx2BidAdapter.js

```diff
@@ -0,0 +1,494 @@
+  },
+
+  /**
+   * Handle data deletion requests
+   * This will both delete local storage data and notify the server
+   */
+  onDataDeletionRequest: function(data) {
+    if (storage.localStorageIsEnabled()) {
+      storage.removeDataFromLocalStorage(USER_ID_KEY);
+    }
+    ajax(USYNC_DELETE_URL, null, null, {
+      method: 'GET',
+      withCredentials: true,
+      browsingTopics: false
+    });
+  },
+};
+
+/**
+ * Creates an outstream renderer for video ads
+ * @param {Bid} bid The bid response
+ * @return {Renderer} A renderer for outstream video
+ */
+function setupOutstreamRenderer(bid) {
+  const renderer = Renderer.install({
+    id: bid.adId,
+    url: OUTSTREAM_PLAYER_URL,
+    loaded: false
+  });
+
+  renderer.setRender(outstreamRender);
+
+  return renderer;
+}
+
+/**
+ * Outstream renderer function called by the renderer
+ * @param {Object} bid The bid object
+ */
+function outstreamRender(bid) {
+  bid.renderer.push(() => {
+    window.ANOutstreamVideo.renderAd({
+      sizes: [bid.width, bid.height],
+      targetId: bid.adUnitCode,
+      adResponse: bid.adResponse
+    });
+  });
+}
+
+
+/* ========
+ * Helpers
+ *=========
+ */
+
+/**
+ * Checks if the bid request has banner media type
+ * @param {BidRequest} bidRequest
+ * @return {boolean} True if has banner media type
+ */
+function hasBannerFormat(bidRequest) {
+  return !!deepAccess(bidRequest, 'mediaTypes.banner');
+}
+
+/**
+ * Checks if the bid request has video media type
+ * @param {BidRequest|BidRequest[]} bidRequest bid request or array of bid requests
+ * @return {boolean} True if has video media type
+ */
+function containsVideoRequest(bidRequest) {
+  if (Array.isArray(bidRequest)) {
+    return bidRequest.some(bid => !!deepAccess(bid, 'mediaTypes.video'));
+  }
+  return !!deepAccess(bidRequest, 'mediaTypes.video');
+}
+
+/**
+ * Validates basic bid parameters
+ * @param {BidRequest} bidRequest
+ * @return {boolean} True if parameters are valid
+ */
+function isParamsValid(bidRequest) {
+  if (!bidRequest.params) {
+    return false;
+  }
+
+  if (bidRequest.params.e2etest) {
+    return true;
+  }
+
+  if (!bidRequest.params.publisher_id) {
+    logError('trustx2 validation failed: Publisher ID (publisher_id) not declared');
+    return false;
+  }
+
+  if (!bidRequest.params.placement_id) {
+    logError('trustx2 validation failed: Placement ID (placement_id) not declared');
+    return false;
+  }
+
+  const hasMediaType = containsVideoRequest(bidRequest) || hasBannerFormat(bidRequest);
+  if (!hasMediaType) {
+    return false;
+  }
+
+  return true;
+}
+
+/**
+ * Validates banner bid request
+ * @param {BidRequest} bidRequest
+ * @return {boolean} True if valid banner bid request
+ */
+function isBannerValid(bidRequest) {
+  // If there's no banner no need to validate
+  if (!hasBannerFormat(bidRequest)) {
+    return true;
+  }
+
+  const banner = deepAccess(bidRequest, 'mediaTypes.banner');
+  if (!Array.isArray(banner.sizes)) {
```

Comment:
what's the deal with all this excessive validation? have you seen an ad unit in the wild lacking sizes?